File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

       int n = s.length();
       HashMap<Character,Integer> map = new HashMap<>();
       int l =0;
       int r = 0;
       int ans = 0;

       while(r<n){
        char ch = s.charAt(r);
        map.put(ch,map.getOrDefault(ch,0)+1);
        
        int length = r-l+1;
        if(length>map.size()){
            char left = s.charAt(l);
            map.put(left,map.get(left)-1);
            if(map.get(left)==0){
                map.remove(left);
            }
            l++;
        }
        
        if(map.size()==length){
            ans = Math.max(ans,length);
        }
        r++;
       } 
       return ans;
    }
}
    

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int>m;
        int j =0;
        int max_length = 0;
        int length = 0;

        for(int i = 0; i < s.size(); i++){
            if(m.find(s[i]) != m.end()){
                j = max(j, m[s[i]] + 1);
            }

            m[s[i]] = i;
            length = i - j + 1;
            if(length > max_length){
                max_length = length;
            }

            
        }
        return max_length;
    }
};